File: src/crewlet/org/hierarchy.py
"""Hierarchy traversal and permission resolution utilities."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from crewlet.org.models import Organization, OrgUnit, Role


def get_manager(role: Role, org: Organization) -> Role | None:
    """Find the manager of a role (the role that lists this role in its manages)."""
    for candidate in org.all_roles():
        if role.name in candidate.manages:
            return candidate
    return None


def get_reports(role: Role, org: Organization) -> list[Role]:
    """Find all direct reports of a role."""
    reports = []
    for name in role.manages:
        found = org.get_role(name)
        if found is not None:
            reports.append(found)
    return reports


def get_ancestors(role: Role, org: Organization) -> list[Role]:
    """Get the management chain from this role up to the top.

    Returns a list starting with the direct manager and ending with
    the topmost manager (who has no manager).
    """
    ancestors = []
    current = role
    seen = {current.name}
    while True:
        manager = get_manager(current, org)
        if manager is None or manager.name in seen:
            break
        ancestors.append(manager)
        seen.add(manager.name)
        current = manager
    return ancestors


def get_unit_for_role(role: Role, org: Organization) -> OrgUnit | None:
    """Find the OrgUnit that directly contains this role."""
    for unit in org.units:
        found = _find_unit_containing_role(unit, role.name)
        if found is not None:
            return found
    return None


def get_unit_chain_for_role(role: Role, org: Organization) -> list[OrgUnit]:
    """Find the chain of OrgUnits from outermost to the unit containing the role.

    Returns ``[grandparent_unit, parent_unit, direct_unit]`` — ordered
    from the top-level structure entry down to the unit that directly
    contains the role. Returns an empty list if the role is not found.
    """
    for unit in org.units:
        chain: list[OrgUnit] = []
        if _build_unit_chain(unit, role.name, chain):
            chain.reverse()  # built innermost-first, reverse to outermost-first
            return chain
    return []


def is_unit_lead(role: Role, org: Organization) -> bool:
    """Check whether *role* is the designated lead of any unit in the org.

    A role can be lead of a parent unit even if it physically lives in
    a child unit (e.g. a department lead whose role is in a team).
    """
    return get_lead_depth(role, org) >= 0


def get_effective_lead(unit: OrgUnit, org: Organization) -> Role | None:
    """Return the effective lead for a unit, including inherited leads.

    When a unit's lead is a direct role or descendant role,
    ``unit.get_lead()`` resolves it.  When the lead was inherited from
    a parent unit (and therefore lives outside this unit's subtree),
    this function falls back to an org-wide role lookup.
    """
    if not unit.lead:
        return None
    # Fast path: lead is in this unit's own roles / descendants.
    lead = unit.get_lead()
    if lead is not None:
        return lead
    # Inherited lead — find the role anywhere in the org.
    return org.get_role(unit.lead)


def is_lead_of_unit(role: Role, unit: OrgUnit) -> bool:
    """Check whether *role* is the designated lead of a specific unit."""
    return unit.lead == role.name


def get_lead_depth(role: Role, org: Organization) -> int:
    """Return the depth at which this role is a lead, or -1 if not a lead.

    Depth 0 = lead of a top-level structure entry (like a division/dept).
    Depth 1 = lead of a child unit. And so on.
    A lower depth means higher authority.
    """
    for unit in org.units:
        depth = _find_lead_depth(unit, role.name, 0)
        if depth >= 0:
            return depth
    return -1


# --- Private helpers ---


def _find_unit_containing_role(unit: OrgUnit, role_name: str) -> OrgUnit | None:
    """Find the unit whose direct ``roles`` list contains *role_name*."""
    for r in unit.roles:
        if r.name == role_name:
            return unit
    for child in unit.children:
        found = _find_unit_containing_role(child, role_name)
        if found is not None:
            return found
    return None


def _build_unit_chain(unit: OrgUnit, role_name: str, chain: list[OrgUnit]) -> bool:
    """Build the chain of units from *unit* down to the one containing *role_name*.

    Appends to *chain* in reverse order (innermost first) during recursion,
    then the caller reverses once at the end. This avoids O(n) inserts.
    Returns True if found.
    """
    # Check if role is directly in this unit
    for r in unit.roles:
        if r.name == role_name:
            chain.append(unit)
            return True
    # Recurse into children
    for child in unit.children:
        if _build_unit_chain(child, role_name, chain):
            chain.append(unit)  # append parent after child (reversed later)
            return True
    return False


def _find_lead_depth(unit: OrgUnit, role_name: str, depth: int) -> int:
    """Find the depth at which *role_name* is a lead. Returns -1 if not found."""
    if unit.lead == role_name:
        return depth
    for child in unit.children:
        result = _find_lead_depth(child, role_name, depth + 1)
        if result >= 0:
            return result
    return -1

File: internal/org/hierarchy.go
package org

import "slices"

// The hierarchy IS the execution graph. Who a seat reports to decides where
// a handoff goes, what a lead's roster contains, and where an escalation
// terminates — so these walks are read on prompt-building and routing
// paths, not just by the dashboard.
//
// They are methods on the org rather than free functions over (role, org)
// because every one of them needs the whole company to answer: management
// is expressed on the MANAGING seat, so finding a seat's manager is a
// search, not a field read.

// Manager returns the seat that lists r's HANDLE in its manages, or nil for
// a top-level seat.
//
// Management is stored on the manager, which is what lets one seat manage
// across units — a root-level CEO managing a VP three levels down needs no
// entry anywhere near that VP. Every entry is a handle by the time this
// reads one: authored entries name a seat by handle, and
// [Organization.expandManages] turns a unit key into the handles in it.
func (o *Organization) Manager(r *Role) *Role {
	handle := r.Handle()
	if handle == "" {
		return nil
	}
	for candidate := range o.AllRoles() {
		if slices.Contains(candidate.Manages, handle) {
			return candidate
		}
	}
	return nil
}

// Reports returns r's direct reports, in the order r lists them. A handle
// that resolves to no seat is skipped — a chart can legitimately be
// half-wired mid-bootstrap.
func (o *Organization) Reports(r *Role) []*Role {
	var out []*Role
	for _, handle := range r.Manages {
		if found := o.Role(handle); found != nil {
			out = append(out, found)
		}
	}
	return out
}

// Ancestors returns the management chain above r, direct manager first and
// the topmost manager last.
//
// The walk stops on a repeat rather than assuming the chart is a tree: a
// config can express a cycle, and a prompt builder that looped forever on
// one would take the whole turn with it. A cycle simply ends the chain.
func (o *Organization) Ancestors(r *Role) []*Role {
	var out []*Role
	seen := map[string]struct{}{r.Handle(): {}}
	current := r
	for {
		manager := o.Manager(current)
		if manager == nil {
			return out
		}
		if _, repeat := seen[manager.Handle()]; repeat {
			return out
		}
		out = append(out, manager)
		seen[manager.Handle()] = struct{}{}
		current = manager
	}
}

// UnitFor returns the unit that holds r as a DIRECT member, or nil for a
// root-level seat.
//
// BY THE SEAT, NOT BY ANY IDENTITY OF IT. A seat is used by pointer
// everywhere (see [Role]), and a stored revision can still hold two seats
// answering to one string: looked up that way, the second seat was answered
// with the first one's unit, so its prompt named a team it is not in.
func (o *Organization) UnitFor(r *Role) *Unit {
	for u := range o.AllUnits() {
		if slices.Contains(u.Roles, r) {
			return u
		}
	}
	return nil
}

// UnitChainFor returns the units from the outermost down to the one holding
// r — [division, department, team] — or nothing for a root-level seat.
//
// The chain is what onboarding walks (a seat reads the Onboarding page of
// every scope above it) and what the engine hashes to decide that a seat
// has MOVED and must onboard again. Found by the seat itself, as
// [Organization.UnitFor] is: by an identity string, the second of two seats
// sharing one was onboarded for the first one's chain, and moving it
// re-onboarded nothing.
func (o *Organization) UnitChainFor(r *Role) []*Unit {
	for _, u := range o.Units {
		if chain := buildUnitChain(u, r); chain != nil {
			slices.Reverse(chain)
			return chain
		}
	}
	return nil
}

// buildUnitChain collects the units from the one holding r back up to u,
// innermost first. Reversing once at the end beats prepending at every level.
func buildUnitChain(u *Unit, r *Role) []*Unit {
	if slices.Contains(u.Roles, r) {
		return []*Unit{u}
	}
	for _, c := range u.Children {
		if chain := buildUnitChain(c, r); chain != nil {
			return append(chain, u)
		}
	}
	return nil
}

// EffectiveLead returns the seat leading u, including a lead inherited from
// an ancestor.
//
// [Unit.LeadRole] answers only for a lead inside u's own subtree; an
// inherited one lives outside it, so this falls back to an org-wide lookup.
// Nil means the unit has no lead, or names one that does not exist yet.
func (o *Organization) EffectiveLead(u *Unit) *Role {
	// NIL-TOLERANT, because the natural way to ask this is
	// EffectiveLead(o.Unit(key)) and a key that names no unit is an ordinary
	// answer rather than a caller's mistake — a stored row can name a unit
	// a later revision removed.
	if u == nil || u.Lead == "" {
		return nil
	}
	if lead := u.LeadRole(); lead != nil {
		return lead
	}
	return o.Role(u.Lead)
}

// IsUnitLead reports whether r leads any unit in the company.
//
// A seat can lead a unit it does not sit in — a department lead whose own
// seat is in one of its teams is the usual shape — so this asks about the
// units, not about where r lives.
func (o *Organization) IsUnitLead(r *Role) bool { return o.LeadDepth(r) >= 0 }

// LeadDepth returns how deep the unit r leads sits, or -1 when r leads
// none. Depth 0 is a top-level unit, 1 its child, and so on: a LOWER depth
// is higher authority, which is what makes it comparable between seats.
//
// When r leads a unit and one of its descendants — a division lead who also
// leads a team inside it — the shallower one wins, because the authority
// that matters is the widest one. Across separate top-level trees the first
// match wins; a seat leading units in two unrelated divisions has no
// meaningful single depth to report.
func (o *Organization) LeadDepth(r *Role) int {
	handle := r.Handle()
	if handle == "" {
		return -1
	}
	for _, u := range o.Units {
		if depth := leadDepth(u, handle, 0); depth >= 0 {
			return depth
		}
	}
	return -1
}

func leadDepth(u *Unit, handle string, depth int) int {
	if u.Lead == handle {
		return depth
	}
	for _, c := range u.Children {
		if found := leadDepth(c, handle, depth+1); found >= 0 {
			return found
		}
	}
	return -1
}
